crates/ethereum/evm/src/execute.rs

```diff
@@ -128,20 +169,23 @@ where
     fn apply_pre_execution_changes(&mut self) -> Result<(), Self::Error> {
         // Set state clear flag if the block is after the Spurious Dragon hardfork.
         let state_clear_flag =
-            self.chain_spec.is_spurious_dragon_active_at_block(self.block.number());
+            self.chain_spec.is_spurious_dragon_active_at_block(self.input.number);
         self.evm.db_mut().set_state_clear_flag(state_clear_flag);
-        self.system_caller.apply_pre_execution_changes(self.block.header(), &mut self.evm)?;
+        self.system_caller
+            .apply_blockhashes_contract_call(self.input.parent_hash, &mut self.evm)?;
+        self.system_caller
+            .apply_beacon_root_contract_call(self.input.parent_beacon_block_root, &mut self.evm)?;
 
         Ok(())
     }
 
     fn execute_transaction(
         &mut self,
         tx: Recovered<&TransactionSigned>,
-    ) -> Result<(), Self::Error> {
+    ) -> Result<u64, Self::Error> {
         // The sum of the transaction's gas limit, Tg, and the gas utilized in this block prior,
         // must be no greater than the block's gasLimit.
-        let block_available_gas = self.block.gas_limit() - self.gas_used;
+        let block_available_gas = self.input.gas_limit - self.gas_used;
```

Comment:
I believe this is safe to do unchecked because we check this against the available gas before increasing gas used